src/Arcus.WebApi.Logging/RequestTrackingMiddleware.cs

```diff
@@ -68,13 +71,28 @@ public async Task Invoke(HttpContext httpContext)
             }
             else
             {
-                var stopwatch = Stopwatch.StartNew();
+                await TrackRequest(httpContext);
+            }
+        }
+
+        private async Task TrackRequest(HttpContext httpContext)
+        {
+            var stopwatch = Stopwatch.StartNew();
+
+            string requestBody = null;
+            if (_options.IncludeRequestBody)
+            {
+                httpContext.Request.EnableBuffering();
+                requestBody = await GetRequestBodyAsync(httpContext);
+            }
 
-                string requestBody = null;
-                if (_options.IncludeRequestBody)
+            Stream originalResponseBodyStream = null;
+            using (Stream temporaryResponseBodyStream = DetermineResponseBodyBuffer())
+            {
+                if (_options.IncludeResponseBody)
                 {
-                    httpContext.Request.EnableBuffering();
-                    requestBody = await GetRequestBodyAsync(httpContext);
+                    originalResponseBodyStream = httpContext.Response.Body;
```

Comment:
Why are we doing this exactly? Because we want to limit the amount of tracked content when specified, and we don't want to influence the actual response or?

Just trying to understand, not critisizm.

I'd move this and everything on https://github.com/arcus-azure/arcus.webapi/pull/203/files#diff-3548ef76495796450e2d323886652ae89240370bb37e6ff76b9c4a7565f0e3c5R111-R115 to a dedicated method for better reading. For example https://github.com/arcus-azure/arcus.webapi/pull/203/files#diff-3548ef76495796450e2d323886652ae89240370bb37e6ff76b9c4a7565f0e3c5R109-R115 can be moved to `CopyTemporaryStreamToOriginalResponseStreamAsync`?

Might be good to add a comment why we are doing this as well.

Reply:
Yeah, just like the request stream, the response stream is not seekable. But unlike the request stream, we don't have a `.EnableBuffering` that makes the body available for multiple reads. So we have to temporary store the response's body in our own seekable stream which we inspect after the processing and track (to app insights or whatever). But after we have tracked the response body (and stream's position is altered), we need to make sure that the remaining middleware components after this one will have the complete response stream. That's why we first reset the position, and then write our temporary stream a second time to the original response stream. 
For the remaining middleware components, it will look like the response was created for the first time.

Makes somewhat sense?
I'll check the links you send to see if we can indeed move something to a separate method, but I would be careful with this since it can make things more complicated.

Reply:
Certainly, thanks for sharing. I'd add this briefly as a comment, personally.

Reply:
Ok, I'll describe the comment there more for future references.